### PR TITLE
Update operator-sdk to 0.17.0; fixup reconcile loop

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Updates operator-sdk to v0.17.0, plus additional fixes. Commits:

- Update operator-sdk (see commit message for detailed changes)
- Update autogenerated files (separate commit to make diff readable)
- Fix issue where we were issuing a `client.update` for each Role and RoleBinding on every reconcile (now do a diff)
- Add missing `patch serviceaccount` permission to controller's role

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16720

### Is it tested? How?
Tested on `crc`, no issues.

### CODE REVIEW

File: deploy/role.yaml [hunk 1]
```diff
@@ -31,6 +31,7 @@ rules:
   - get
   - create
   - update
+  - patch
```
Comment (sleshchenko) [inline deploy/role.yaml:34]: hopefully we already have `oc apply -f ./deploy/role.yaml` ))
Screenshot_20200423_113726

**Comment by amisevsk** [comment]
@sleshchenko Yeah, there are a few issues that I'm not 100% clear on:
- The openapi rules / validation seems very confusing, e.g. I believe `listMapKeys` are required when `listType=map`. This also appears to depend on the version of the sdk and the version of k8s on the cluster. 
- I didn't initially have the `patch` rbac issue, but then it popped up :shrug:

**Comment by davidfestal** [comment]
@amisevsk About `listMapKey` some explanation is there:
https://github.com/kubernetes/kube-openapi/blob/master/pkg/idl/doc.go#L63

Examles of `listMapKey` and the relationship with `patchMergeKey` can be found in this K8S API example:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2166

**Comment by amisevsk** [comment]
@davidfestal The issue is mostly that I can't get the annotations to work as expected:
```bash
# add 	// +listType=map +listMapKey=name
$ operator-sdk generate crds
INFO[0000] Running CRD generator.                       
/home/amisevsk/git/che-workspace-operator/pkg/apis/workspace/v1alpha1/component.go:32:2: ListType must be either "map", "set" or "atomic"
FATA[0000] error generating CRDs from APIs in pkg/apis: error generating CRD manifests: error running CRD generator: not all generators ran successfully
```

```bash
# add 	// +listType="map" +listMapKey=name
operator-sdk generate crds
INFO[0000] Running CRD generator.                       
/home/amisevsk/git/che-workspace-operator/pkg/apis/workspace/v1alpha1/component.go:31:2: extra arguments provided: "+listMapKey=name" (at <input>:1:7)
/home/amisevsk/git/che-workspace-operator/pkg/apis/workspace/v1alpha1/component.go:31:2: extra arguments provided: "+listMapKey=name" (at <input>:1:7)
FATA[0000] error generating CRDs from APIs in pkg/apis: error generating CRD manifests: error running CRD generator: not all generators ran successfully
```
This is with operator-sdk `v0.17.0`

File: deploy/role.yaml (hunk 1 above)
Comment (amisevsk) [reply deploy/role.yaml:34]: Finally figured this out: we had conflicting definitions of the `che-workspace-controller` clusterrole. `kubectl apply`ing the entire folder would create the role we meant, and then overwrite it. 

I've reworked a the role/rolebinding definitions to a single yaml file now.

**Comment by amisevsk** [comment]
Added commit https://github.com/che-incubator/che-workspace-operator/pull/57/commits/1cf2b19407570dd6f989f96654a5b5b05a03e437 to clean up our RBAC definitions for the controller and retested on `crc`.

Tried to test on latest minikube (k8s 1.18) and ran into https://github.com/eclipse/che/issues/16791

**Comment by amisevsk** [comment]
Was able to test on minikube (k8s 1.17) as well. Fixed one minor issue (that should be present in master as well?): https://github.com/che-incubator/che-workspace-operator/pull/57/commits/211e87e191ca0cf062b9b4ee37fbcc245ef67d5d

**Comment by davidfestal** [comment]
> @davidfestal The issue is mostly that I can't get the annotations to work as expected:
> 
> ```shell
> # add 	// +listType=map +listMapKey=name
> $ operator-sdk generate crds
> INFO[0000] Running CRD generator.                       
> /home/amisevsk/git/che-workspace-operator/pkg/apis/workspace/v1alpha1/component.go:32:2: ListType must be either "map", "set" or "atomic"
> FATA[0000] error generating CRDs from APIs in pkg/apis: error generating CRD manifests: error running CRD generator: not all generators ran successfully
> ```
> 
> ```shell
> # add 	// +listType="map" +listMapKey=name
> operator-sdk generate crds
> INFO[0000] Running CRD generator.                       
> /home/amisevsk/git/che-workspace-operator/pkg/apis/workspace/v1alpha1/component.go:31:2: extra arguments provided: "+listMapKey=name" (at <input>:1:7)
> /home/amisevsk/git/che-workspace-operator/pkg/apis/workspace/v1alpha1/component.go:31:2: extra arguments provided: "+listMapKey=name" (at <input>:1:7)
> FATA[0000] error generating CRDs from APIs in pkg/apis: error generating CRD manifests: error running CRD generator: not all generators ran successfully
> ```
> 
> This is with operator-sdk `v0.17.0`

@amisevsk did you finally succeed in setting these annotations ?

**Comment by amisevsk** [comment]
@davidfestal No, for now we do not apply the annotations where it was throwing errors. We also have an issue on Kubernetes 1.8 due to embedding pods in our status fields: https://github.com/eclipse/che/issues/16791

**Pull request merged**